curp/src/members.rs

```diff
@@ -1,134 +1,172 @@
-                id, all_members
-            )
-        });
-        Self {
-            self_id: id,
-            self_address: address,
-            peers: all_members,
+    pub fn new(all_members: HashMap<String, String>, self_name: &str) -> Self {
+        let mut member_id = 0;
+        let mut members = HashMap::new();
+        for (name, address) in all_members {
+            let id = Self::calculate_member_id(&address, "", None);
+            if name == self_name {
+                member_id = id;
+            }
+            let member = Member { id, name, address };
+            let _ig = members.insert(id, member);
         }
+        debug_assert!(member_id != 0, "self_id should not be 0");
+        let mut cluster_info = Self {
+            cluster_id: 0,
+            member_id,
+            members,
+        };
+        cluster_info.gen_cluster_id();
+        cluster_info
     }
 
-    /// get server address via server id
+    /// Get server address via server id
     #[must_use]
     #[inline]
-    pub fn address(&self, id: &str) -> Option<&str> {
-        if id == self.self_id {
-            Some(self.self_address())
-        } else {
-            self.peers.get(id).map(String::as_str)
-        }
+    pub fn address(&self, id: ServerId) -> Option<&str> {
+        self.members
+            .values()
+            .find(|t| t.id == id)
+            .map(|t| t.address.as_str())
+    }
+
+    /// Get the current member
+    #[allow(clippy::indexing_slicing)] // self member id must be in members
+    fn self_member(&self) -> &Member {
+        &self.members[&self.member_id]
     }
 
-    /// get the current server address
+    /// Get the current server address
     #[must_use]
     #[inline]
     pub fn self_address(&self) -> &str {
-        self.self_address.as_str()
+        &self.self_member().address
     }
 
-    /// get the current server id
+    /// Get the current server id
     #[must_use]
     #[inline]
-    pub fn self_id(&self) -> &ServerId {
-        &self.self_id
+    pub fn self_name(&self) -> &str {
+        &self.self_member().name
     }
 
-    /// get peers id
+    /// Get peers id
     #[must_use]
     #[inline]
-    pub fn peers_id(&self) -> Vec<ServerId> {
-        self.peers.keys().cloned().collect()
+    pub fn peers_ids(&self) -> Vec<ServerId> {
+        self.members
+            .values()
+            .filter(|t| t.id != self.member_id)
+            .map(|t| t.id)
+            .collect()
     }
 
-    /// calculate the member id
-    #[must_use]
-    #[inline]
-    pub fn gen_member_id(&self, cluster_name: &str) -> u64 {
-        let ts = SystemTime::now()
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .unwrap_or_else(|e| unreachable!("SystemTime before UNIX EPOCH! {e}"))
-            .as_secs();
+    /// Calculate the member id
+    fn calculate_member_id(address: &str, cluster_name: &str, timestamp: Option<u64>) -> ServerId {
```

Comment:
yes, A timestamp is not required fo computing the member ID when initializing the cluster, because to ensure that all nodes calculate the same ID, New nodes added need to be calculated based on the timestamp